### PR TITLE
Add support to CLI build

### PR DESCRIPTION
**Pull request by ulyssessouza**
This add support for building with the CLI by setting the env var `COMPOSE_DOCKER_CLI_BUILD=1`.
This feature can be used both by `docker-compose build` and `docker-compose up --build`.

Note that this feature is EXPERIMENTAL, therefore a warning message is printed on every execution

Signed-off-by: Nao YONASHIRO <yonashiro@r.recruit.co.jp>
Signed-off-by: Ulysses Souza <ulysses.souza@docker.com>

This PR is a squashed and resumed version of https://github.com/docker/compose/pull/6584

As a followup we need to check following points in other PRs
- [x] Add `--progress` flag in `docker-compose build` to control the format of the output
- [x] Add a warning if someone has this feature enabled, and uses the --compress or --parallel flag
- [x] Add integration tests for `build` and `up --build`

### CODE REVIEW

**Comment by abhaychrungoo** [comment]
I see a couple of issues. Both are likely related and could be summarized as  
- Not tagging the final image(s) correctly

Case 1: 
On first run with  build --up, the images are built twice. Once with buildkit, and once without

**Comment by orisano** [comment]
`build --up`? Is it `up --build`?

**Comment by abhaychrungoo** [comment]
@orisano :  up --build. 
@ulyssessouza : What warning are you expecting to provide with the --parallel flag ?

**Comment by orisano** [comment]
Maybe resolve that
https://github.com/orisano/compose/commit/ee60803fad7d62b275ec4235dad3853aa9a3a3e0

**Comment by abhaychrungoo** [comment]
I will test this again and provide feedback. However I do feel that in the broader scheme of things, we should be adding something to the integration test suite in this PR.

Not sure of this is covered in any any testcases already.
- the  tagged image sha matches the sha of the final (or target stage) image. 

I'm happy to dig into the code, if that helps to move this PR along.

**Comment by abhaychrungoo** [comment]
> orisano@ee60803

Yes, it fixes the issue for both ```build``` and ```up --build```

**Comment by orisano** [comment]
https://github.com/orisano/compose/commit/496562749f743708cb95104b2e26f15f968ef9f5

I implemented a prototype of `--progress` flag on `docker-compose build`.

**Comment by orisano** [comment]
https://github.com/orisano/compose/commit/a633c9a16763cfb4c0446c2a2eb5dbcb1234ab67

**Comment by orisano** [comment]
@ulyssessouza What could I do?

**Comment by ulyssessouza** [comment]
Thanks a lot @orisano ! 
Today I'll integrate this 3 commits, maybe rephrase the warnings and add integration tests to all of this.

Then I think the scope of this PR is closed.

**Comment by orisano** [comment]
Please add this commit.
https://github.com/orisano/compose/commit/8e1a715d027e0be7dba8874ed22b13d22cd36065

Maybe, it is the mistake of conflict resolves.

**Comment by ulyssessouza** [comment]
@orisano Fixed

**Comment by orisano** [comment]
@ulyssessouza thanks!

File: tests/integration/service_test.py [hunk 1]
```diff
@@ -966,6 +968,43 @@ def test_build(self):
...
+        with open(os.path.join(base_dir, 'Dockerfile'), 'w') as f:
+            f.write("FROM busybox\n")
+
+        service = self.create_service('web',
+                                      build={'context': base_dir},
+                                      environment={
+                                          'COMPOSE_NATIVE_BUILDER': '1',
+                                          'DOCKER_BUILDKIT': '1',
+                                      })
+        service.build(cli=False)
```
Comment (orisano) [inline tests/integration/service_test.py]: `cli=False`? Is it maybe `cli=True`?

Comment (ulyssessouza) [reply tests/integration/service_test.py]: Fixed

**Comment by orisano** [comment]
@ulyssessouza I fixed a failed test on python2.
https://github.com/orisano/compose/commit/19368476c5686d51a1198da66b693e3804b22399

File: compose/cli/main.py [hunk 2]
```diff
@@ -263,14 +263,17 @@ def build(self, options):
         Usage: build [options] [--build-arg key=val...] [SERVICE...]
 
         Options:
+            --build-arg key=val     Set build-time variables for services.
             --compress              Compress the build context using gzip.
             --force-rm              Always remove intermediate containers.
+            -m, --memory MEM        Sets memory limit for the build container.
```
Comment (ndeloof) [inline compose/cli/main.py]: Heterogeneous Set/Sets

File: compose/service.py [hunk 3]
```diff
@@ -1701,3 +1709,132 @@ def rewrite_build_path(path):
...
+                    appear = True
+                yield json.dumps({"stream": line})
+
+        with open(iidfile) as f:
+            line = f.readline()
+            image_id = line.split(":")[1].strip()
+        os.remove(iidfile)
+
+        if not appear:
+            yield json.dumps({"stream": "{}{}\n".format(magic_word, image_id)})
```
Comment (ndeloof) [inline compose/service.py]: Not sure to understand what this is for.
If a workaround/hack on cli weird behaviour, would better be documented with a few comments.

Comment (ulyssessouza) [reply compose/service.py]: Sure! 
This is related to the way Service::build() gets the `image_id` after building (https://github.com/docker/compose/blob/master/compose/service.py#L1110).
In the case of a build with buildkit enabled, we do not have this line in the logs. So it's added `manually`.
I'll add a comment on it.

File: compose/cli/main.py (hunk 2 above)
Comment (ulyssessouza) [reply compose/cli/main.py]: Actually, this is a pre-existing `typo`, but I'll remove the 's'

**Comment by ulyssessouza** [comment]
We decided to rename the envvar from `COMPOSE_NATIVE_BUILDER` to `COMPOSE_DOCKER_CLI_BUILD`. 
This change makes it explicit 3 things:
- The envvar is of scope os `docker-compose` by starting with `COMPOSE`
- What is used? The `docker cli`
- To do what (the scope of the action)? The build

**Comment by ulyssessouza** [comment]
Thanks @orisano for the contribution! This feature will be available in the next RC.
Please feel free to reach us out in the #docker-compose channel in the docker community slack

**Comment by thaJeztah** [comment]
Thanks everyone! Looking forward to see people use this feature 👍 (be sure to post feedback!)

**Comment by itaymelamed** [comment]
When it would be possible to start using it in docker compose?

**Comment by ulyssessouza** [comment]
@itaymelamed In the next release candidate (1.25.0-rc3). It has no specific date but it's comming very soon (weeks).

**Comment by abhaychrungoo** [comment]
We've used this in our CI fairly extensively over the past few weeks.
Works like a charm for our usecases. No issues reported.

Thanks @ulyssessouza

**Comment by thaJeztah** [comment]
Thanks for the feedback, @abhaychrungoo ! That's good to hear 🤗

**Comment by bkaidbb** [comment]
First of all, it works great and getting a lot faster builds as a result.  Thanks! 

I installed latest via:
```
curl -L https://dl.bintray.com/docker-compose/master/docker-compose-`uname -s`-`uname -m` -o /usr/local/bin/docker-compose
chmod +x /usr/local/bin/docker-compose
docker-compose --version
```
Result: `docker-compose version 1.25.0dev, build bc57a1bd`

I noticed a difference now in behavior with `restart: on-failure` that isn't ideal.  Previously if I did docker-compose up and the service restarted (in my case the database dependency hadn't finished starting), the restart would happen and the command would continue. Now if I do `DOCKER_BUILDKIT=1 COMPOSE_DOCKER_CLI_BUILD=1 docker-compose up --build myservice` the restart happens but docker-compose exits and doesn't reattach to the restarted instance.

**Comment by bkaid** [comment]
First of all, it works great and getting a lot faster builds as a result.  Thanks! 

I installed latest via:
```
curl -L https://dl.bintray.com/docker-compose/master/docker-compose-`uname -s`-`uname -m` -o /usr/local/bin/docker-compose
chmod +x /usr/local/bin/docker-compose
docker-compose --version
```
Result: `docker-compose version 1.25.0dev, build bc57a1bd`

I noticed a difference now in behavior with `restart: on-failure` that isn't ideal.  Previously if I did docker-compose up and the service restarted (in my case the database dependency hadn't finished starting), the restart would happen and the command would continue. Now if I do `DOCKER_BUILDKIT=1 COMPOSE_DOCKER_CLI_BUILD=1 docker-compose up --build myservice` the restart happens but docker-compose exits and doesn't reattach to the restarted instance.

**Comment by thaJeztah** [comment]
@bkaid thanks for reporting: would you be able to open a new issue and provide a minimum reproduction example?

**Comment by bkaid** [comment]
@thaJeztah Sorry - disregard.  I tried on latest release `version 1.24.1, build 4667896b` and it works the same way so I don't think there is a change with this new version and I think I just misunderstood how it worked.

docker-compose.yml:
```
version: "3.7"

services:
  rabbitmq:
    image: rabbitmq:3.7.15-management-alpine

  myservice:
    image: alpine:3.10.2
    restart: on-failure
    depends_on:
      - rabbitmq
    entrypoint: sh -c "echo Starting up... && sleep 10 && echo Exiting on error && exit 1"
```

This will restart my service and keep showing updated logs and not exit:
`docker-compose up rabbitmq myservice`

This will restart my service in the background and exit:
`docker-compose up myservice`

**Comment by lig** [comment]
@ulyssessouza no pressure but it starting to become months :(

**Comment by smamessier** [comment]
I'd like to use `--ssh default` to forward the SSH agent at build time using compose. Is there a way we can forward all options to `docker build` CLI?

**Comment by TortelliniLeap** [comment]
@smamessier Also looking to do the same, here.

**Comment by TortelliniLeap** [comment]
@smamessier I'm interested in support for this as well.

**Comment by chris-crone** [comment]
This PR is included in the 1.25.0-rc3 release that you can find [here](https://github.com/docker/compose/releases/tag/1.25.0-rc3). It will also be part of the next Docker Desktop edge release.

**Comment by aleygues** [comment]
@smamessier I'm also interested in this feature!

**Comment by lig** [comment]
@chris-crone it would be nice to see this mentioned in the release docs. It looks like a pretty significant new feature.

**Comment by pasqoo** [comment]
How do I execute the equivalent of `docker build --ssh default=/home/user/.ssh/keyfile .` with _docker-compose_?
I have tried `COMPOSE_DOCKER_CLI_BUILD=1 docker-compose build` but it's not working.
I guess that only handles connections to an _ssh-agent_, instead of a file as in this case (`/home/user/.ssh/keyfile`).

**Comment by thaJeztah** [comment]
@pasqoo not supported yet, but there's a PR working on that to add it to the compose-file syntax; https://github.com/docker/compose/pull/7046. It won't be added as a command-line flag, because you want to control which builds have access to your ssh-agent (as the compose file may have multiple builds, and only some of them could need access)

**Comment by shabeerak** [comment]
@thaJeztah do you know when **--ssh** feature will be released?
